#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd


In [2]:
data_path = '../data/supermarket_sales.csv'
df = pd.read_csv(data_path)

In [3]:
type(df)

pandas.core.frame.DataFrame

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [4]:
df.groupby('Branch')['gross income'].sum()

Branch
A    5057.1605
B    5057.0320
C    5265.1765
Name: gross income, dtype: float64

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [5]:
df.groupby('City')['Branch'].unique()

City
Mandalay     [B]
Naypyitaw    [C]
Yangon       [A]
Name: Branch, dtype: object

In [6]:
df.groupby('City')[['Total','cogs','Quantity', 'gross income']].agg(['sum','mean'])

Total                   cogs             Quantity            \
                   sum        mean        sum        mean      sum      mean   
City                                                                           
Mandalay   106197.6720  319.872506  101140.64  304.640482     1820  5.481928   
Naypyitaw  110568.7065  337.099715  105303.53  321.047348     1831  5.582317   
Yangon     106200.3705  312.354031  101143.21  297.480029     1859  5.467647   

          gross income             
                   sum       mean  
City                               
Mandalay     5057.0320  15.232024  
Naypyitaw    5265.1765  16.052367  
Yangon       5057.1605  14.874001

Філіал С з найвищим прибутком і складає 5265.18. Цей філіал знаходиться в місті Naypyitaw. 
Для прийняття рішення про відкриття ще одного магазину в цьому місті було проаналізовано наступні показники: загальна сума продажіва, собівартість проданої продукції, к-ть покупок та валовий дохід. 
  Дослідивши дані, бачимо, що місто Naypyitaw має найкращі показники серед усіх міст:
- Найвищий середній валовий дохід — 16,05
- Найбільша середня кількість товарів у чеку — 5,58
- Найвищий загальний обсяг продажу — 110 568,7
Однак, місто Naypyitaw  має найвищу середню собівартість товарів, тому було додатково проаналізовано середній відсоток валового прибутку.
Результати показали, що відсоток валового прибутку є однаковим для всіх міст, що свідчить про стабільну цінову політику та маржу прибутковості.
Отже,  можна зробити висновок що місто є перспективним для відкриття ще одного магазину, бо споживачі активно купують та готові витрачати більше.

In [7]:
df.groupby('City')['gross margin percentage'].mean()

City
Mandalay     4.761905
Naypyitaw    4.761905
Yangon       4.761905
Name: gross margin percentage, dtype: float64

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [8]:
invoices_by_category = pd.pivot_table(df, index='Branch', columns='Product line', values='Quantity', aggfunc='count')
invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [9]:
invoices_by_category.groupby('Branch')['Electronic accessories'].max()

Branch
A    60
B    55
C    55
Name: Electronic accessories, dtype: int64

4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


**Гіпотеза 1 Які товари найбільше і найменше купують в розрізі годин та днів тижня**  


In [10]:
df['Hour'] = df['Time' ].apply(lambda x: int(x.split(':')[0]))


In [11]:
res1 = pd.pivot_table(df, index='Product line', columns='Hour', values='Quantity', aggfunc='sum')
res1

Hour                     10   11   12   13   14   15  16   17   18   19   20
Product line                                                                
Electronic accessories   89   98   89   76   84   91  62   84  103  105   90
Fashion accessories      90   67   91  112   86   57  88   56   98  100   57
Food and beverages       90   89   63   67   78  124  74   35   79  138  115
Health and beauty        65   65   83  109  119   64  49   63   79   87   71
Home and lifestyle      100  113   71  111   49   99  51  123   59   88   47
Sports and travel        91   81  104  110   79   95  96   54   57  131   22

In [12]:
df['Date'] = pd.to_datetime(df['Date'])
df['Day'] = df['Date'].dt.day_name()

In [13]:
res2 = pd.pivot_table(df, index='Product line', columns='Day', values='Quantity', aggfunc='sum')
res2

Day                     Friday  Monday  Saturday  Sunday  Thursday  Tuesday  \
Product line                                                                  
Electronic accessories     101     124       159     134       213      151   
Fashion accessories        156      92       179     103       119      119   
Food and beverages          95      98       151     139       116      168   
Health and beauty          145     140       107      80        96      155   
Home and lifestyle         131     111       149     187        95      136   
Sports and travel          130      73       174     135       116      133   

Day                     Wednesday  
Product line                       
Electronic accessories         89  
Fashion accessories           134  
Food and beverages            185  
Health and beauty             131  
Home and lifestyle            102  
Sports and travel             159

Найбільше купують:
13 -15  годин  - Fashion та Health
17 -20 години - Electronic, Food, Sports та Home
Найменше купують:
16-18 годині - Electronic, Health, Fashion та Food
з 20 години - Home та Sports

Найактивніші дні тижня:
Fashion та Sports найбільше купують у суботу.
Food — у середу.
Home  — у неділю.
Найменше купують у: 
понеділок -Fashion та Sports 
середа, четверг - Electronic та Home 
п'ятницю - Food 
неділю - Health

**Рекомендації:**  
1. Забезпечити присутність персоналу у пікові години та пікові дні( щоб було достатньо касирів, наявність консультантів)
2. Розробити систему знижок в той час і в той день коли найменше продається товар( на приклад на товари з категорії Sports зробити знижки у понеділок)
3. У пікові години та дні покупцям можна пропонувати нові товари, або просувати товари з більш дорожчого сегменту.
4. Спланувати логістику таким чином щоб популярні товари були в достатній к-ть в пікові дні та години

**Гіпотеза 2 Дослідити скільки і на які товари витрачають чоловіки та жінки**

In [14]:
res4 = pd.pivot_table(df, index='Product line', columns='Gender', values='Total', aggfunc='sum')
res4

Gender                      Female        Male
Product line                                  
Electronic accessories  27102.0225  27235.5090
Fashion accessories     30437.4000  23868.4950
Food and beverages      33170.9175  22973.9265
Health and beauty       18560.9865  30632.7525
Home and lifestyle      30036.8775  23825.0355
Sports and travel       28574.7210  26548.1055

Ми бачимо що в більшості категорій товарів(Fashion,Food,Home та Sports) жінки витрачають більше ніж чоловіки. В категорії Health and beauty більше витрачають чоловіки(що для мене було нетипово, напевно в цій категорії товарів більше або привабливіше для чоловіків)та Electronic accessories (незначна перевага).  

**Рекомендації:**  
У категоріях ***Fashion,Food,Home та Sports*** орієнтуватися на жіночу аудиторію( рекламні акції, особисті розсилки), розглянути щоб розширити асортимент жіночого товару в цих категоріях.  

У категорії ***Health and beauty*** орієнтуватися на чоловіків. Можливо розглянути розширення товарів і для жінок, бо зазвичай жінки витрачають більше на доглядові засоби.  

Оскільки в категорії ***Electronic accessories*** приблизно однакові витрати, тому варто орієнтуватися як на жінок так і на чоловіків.


**Гіпотеза 3 Хто є наш покупець( з карткою магазина чи звичайний)**

In [15]:
res5 = pd.pivot_table(df, index='Product line', columns='Customer type', values='Total', aggfunc='sum')
res5

Customer type               Member      Normal
Product line                                  
Electronic accessories  24498.4950  29839.0365
Fashion accessories     26323.9620  27981.9330
Food and beverages      31357.6200  24787.2240
Health and beauty       25831.0395  23362.6995
Home and lifestyle      27978.0270  25883.8860
Sports and travel       28234.3005  26888.5260

In [16]:
res6 = pd.pivot_table(df, index='Product line', columns='Customer type', values='Quantity', aggfunc='sum')
res6

Customer type           Member  Normal
Product line                          
Electronic accessories     429     542
Fashion accessories        439     463
Food and beverages         506     446
Health and beauty          428     426
Home and lifestyle         490     421
Sports and travel          493     427

У більшості категорій товарів (Food, Health, Home, Sports) члени клубу купують частіше і витрачають більше. Але у 2 категоріях — Electronic та Fashion — переважають звичайні покупці, як за кількістю, так і за сумою витрат.

**Рекомендації:**  
1.Адаптувати програму лояльності або маркетинг у тих категоріях де звичайні покупці переважають.  
2.Розробити акції та пропозиції для членів клубу у категоріях Electronic та Fashion
